### ISSUE TITLE
Extract PDF contents in to CSV format

### ISSUE THREAD

**Issue by hemalbavisi** [tabulapdf/tabula#466]
Hello,

Thanks for developing such a wonderful tool and I believe it helps a community to large extent. I tried importing the documents I have but I am not getting expected results. Take any specification e.g. AUTOSAR SWS COM from the below mentioned path and I tried to convert the contents into csv format. To some extent it does the job but still output is not upto the mark (particularly table contents) 

http://www.autosar.org/specifications/release-42/software-architecture/communication-stack/

**Comment by jeremybmerrill** [comment]
which tables in particular are causing you trouble? What does the output look like, and what do you expect it to look like?

**Comment by hemalbavisi** [comment]
Please refer below table :

![image](https://cloud.githubusercontent.com/assets/9947012/13176062/267fb86c-d735-11e5-84ae-754dfef5dd31.png)

I am expecting output should come in excel as shown in the table i.e. output should be divided into two separate column - one representing first column of the table e.g. Service name and second representing the second column of the table e.g. Com_SendSignal. Moreever Second column needs to be spitted into two parts wherever necessary e.g Parameters in raw. 

This is how it appears in the output window by the tool

![image](https://cloud.githubusercontent.com/assets/9947012/13176192/22659e12-d736-11e5-8d3f-5161bd146572.png)

**Comment by hemalbavisi** [comment]
Look another example of the same document

![image](https://cloud.githubusercontent.com/assets/9947012/13176291/fa07c3cc-d736-11e5-9e23-cf61f194da4b.png)

and the output

![image](https://cloud.githubusercontent.com/assets/9947012/13176316/34488864-d737-11e5-9fbe-57a96f95839e.png)

It looks more uglier than the previous one 

I am expecting the output should appear as similar to one available in the PDF -> May be my expectation is too much;)

**Comment by jeremybmerrill** [comment]
the master branch (which is not yet available as a pre-built executable/jar) does a better job in Lattice mode (see attached). 

It's not perfect because the proper way to treat tables with spanning cells in a CSV-oriented environment isn't intuitively obvious. We don't export directly to Excel, Excel's odd features like this one aren't supported and won't ever be. I'd posit there's not a good solution at all, especially for the vertical spanning cells case you've got for `Parameters (in)` and  `Value Configuration Class`. (@jazzido, this is a useful test case for https://github.com/tabulapdf/tabula-java/issues/55 )

[tabula-AUTOSAR_SWS_COM_00851.csv.txt](https://github.com/tabulapdf/tabula/files/138270/tabula-AUTOSAR_SWS_COM_00851.csv.txt)

[tabula-AUTOSAR_SWS_COM.csv.txt](https://github.com/tabulapdf/tabula/files/138267/tabula-AUTOSAR_SWS_COM.csv.txt)

**Comment by jeremybmerrill** [comment]
Also, I'd recommend not including the header (i.e. the `SWS_Com_00851` in the first table) in your selection.

**Comment by wbuchanan** [comment]
@jeremybmerrill Would a potential solution be to implement a rule that the top left cell is always the cell that would be filled in cases of spans so the resulting out put would look something like:

```
[SWS_Com_00851]        |   (empty)         |   (empty)      
-----------------------|-------------------|------------------
Service Name:          |   cell contents   |   (empty)    
Syntax                 |   cell contents   |   (empty)    
Service ID (hex)       |   cell contents   |   (empty)    
Sync/Async             |   cell contents   |   (empty)       
Reentrancy             |   cell contents   |   (empty)    
Parameters (in):       |   cell contents   |   cell contents   
(empty)                |   cell contents   |   cell contents   
Parameters (in-out):   |   cell contents   |   (empty)   
Parameters (out):      |   cell contents   |   (empty)    
Return value:          |   cell contents   |   cell contents   
Description            |   cell contents   |   (empty)    
```

Where (empty) is a cell that would have a zero length string and cell contents would have the corresponding contents for the cell (based on the first example provided by @hemalbavisi
![](https://cloud.githubusercontent.com/assets/9947012/13176062/267fb86c-d735-11e5-84ae-754dfef5dd31.png)).  It isn't a perfect solution for what @hemalbavisi asked for, but I think this would make the tool immensely more useful to the broader community who would then only need to spend minimal effort to get things into their ideal format.

**Comment by jeremybmerrill** [comment]
Yes, absolutely that's a good description of intended behavior. The hard part is implementing it. :) If you're interested in taking a stab at it, let me know and I can point you to the relevant parts of the code. (Java, in [tabula-java](https://github.com/tabula/tabula-java), if it matters.)

**Comment by wbuchanan** [comment]
@jeremybmerrill I've actually been trying to figure out some of the extraction stuff on my own for a while so I could integrate tabula-java with [Stata](http://www.stata.com).  In my particular case, I ran into issues with a sporadic multirow column header in a document and wanted to figure out how I could potentially figure out how to keep the values of what are essentially a single cell in a single object.  If there are specific places you think I should look at, by all means let me know and I'll see what I can figure out (if anything).

**Comment by jeremybmerrill** [comment]
@wbuchanan: this is the SpreadsheetExtractionAlgorithm https://github.com/tabulapdf/tabula-java/blob/master/src/main/java/technology/tabula/extractors/SpreadsheetExtractionAlgorithm.java, aka Lattice Mode. That's what parses the Excel-style fancy tables.

**Comment by wbuchanan** [comment]
@jeremybmerrill I've read through the source a few times, but it was fairly difficult for me to follow (I'm still relatively new to Java).  Is there any analogous object to a row?  Perhaps making the Table class behave more like a storage container for rows and/or columns would make it easier to handle these types of cases (e.g., define the dimensions of the tables based on the maximum number of elements in any row/column)?

**Comment by jeremybmerrill** [comment]
Yes, it's kind of dense. This is the algorithm that's being implemented: http://dspace.cc.tut.fi/dpub/bitstream/handle/123456789/21520/Nurminen.pdf?sequence=3

The problem, though, is that we don't yet have a table structure. All we have are a bunch of lines, some vertical, some horizontal.
